packages/loopring_v3/contracts/iface/IExchangeV3.sol

```diff
@@ -889,84 +778,110 @@ contract IExchangeV3 is IExchange
-    ///      requests were processed by the operator.
-    ///
-    ///      Needs to be called by the operator after submitting a block processing
-    ///      withdrawal requests (either onchain or offchain requests) after the block
-    ///      is finalized and before the block is MAX_TIME_TO_DISTRIBUTE_WITHDRAWALS seconds old.
-    ///
-    ///      If the operator fails to do so anyone will be able to call this function
-    ///      and the stake of the exchange will be used to reward the caller of this function.
-    ///      The amount of staked LRC withdrawn is calculated as follows:
-    ///
-    ///      totalFine = withdrawalFineLRC * numWithdrawalRequestsInBlock
-    ///      The caller of the function will be rewarded half this amount,
-    ///      the other half is burned.
-    ///
-    ///      Only withdrawals processed in finalized blocks can be distributed.
-    ///
-    ///      The withdrawals can be done in multiple transactions because the token transfers
-    ///      are more expensive than committing and proving a block, so it's possible more
-    ///      withdrawals requests are processed in a block than can be distributed
-    ///      in an Ethereum block.
-    ///      This function will automatically stop distributing the withdrawals when the amount
-    ///      of gas left is less than MIN_GAS_TO_DISTRIBUTE_WITHDRAWALS.
-    ///      So there are 2 ways to  limit the number of withdrawals:
-    ///          - using the maxNumWithdrawals parameter
-    ///          - limiting the amount of gas in the transaction
-    ///
-    /// @param  blockIdx The block index to distribute the funds from the withdrawal requests for
-    /// @param  maxNumWithdrawals The max number of withdrawals to distribute. Can be lower than the
-    ///         number of withdrawal requests processed in the block. Withdrawals are distributed
-    ///         in the same order the withdrawal requests were processed in the block.
-    ///         If the withdrawals are done in multiple parts we always start from the
-    ///         first withdrawal that was not yet distributed.
-    function distributeWithdrawals(
-        uint blockIdx,
-        uint maxNumWithdrawals
+        view
+        returns (uint);
+
+    /// @dev Allows an agent to transfer ERC-20 tokens for a user using the allowance
+    ///      the user has set for the exchange. This way the user only needs to approve a single exchange contract
+    ///      for all exchange/agent features, which allows for a more seamless user experience.
+    ///
+    ///      This function can only be called by an agent.
+    ///
+    /// @param from The address of the account that sends the tokens.
+    /// @param to The address to which 'amount' tokens are transferred.
+    /// @param token The address of the token to transfer (ETH is and cannot be suppported).
+    /// @param amount The amount of tokens transferred.
+    function onchainTransferFrom(
```

Comment:
I think this is mainly for fast withdrawal to move one user's layer1 balance to another user's layer1 balance. So this is required.

Reply:
It's not really needed, but otherwise people would have to approve tokens on potentially lots of Agent contracts, so I think it is worth it for the user experience.

Also, in most cases (at least for general extended exchange functionality) the agent should be a contract that cannot be updated with limited functionality so it can only call the functions it needs with some strict rules.

Reply:
I would allow EOAs because I could see how that could be useful if people want to access the account from multiple addresses. I don't think it would increase security because the attacker would simply use a contract as well.